core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,270 @@
+ * on random URLs to support resource URLs (second domain).
+ *
+ * @see ResourceDomainConfiguration
+ * @see ResourceDomainFilter
+ *
+ * @since TODO
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class ResourceDomainRootAction implements UnprotectedRootAction {
+
+    private static final Logger LOGGER = Logger.getLogger(ResourceDomainRootAction.class.getName());
+
+    @CheckForNull
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getUrlName() {
+        return "static-files";
+    }
+
+    public void doIndex(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+        } else {
+            rsp.sendError(404, "Jenkins serves only static files on this domain.");
+        }
+    }
+
+    public Object getDynamic(String id, StaplerRequest req, StaplerResponse rsp) throws Exception {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+            return null;
+        }
+
+        Token token = Token.decode(id);
+        if (token == null) {
+            rsp.sendError(404, "Jenkins serves only static files on this domain.");
+            return null;
+        }
+
+        String authenticationName = token.username;
+        String browserUrl = token.path;
+
+        long creationDate = token.timestamp;
+        long age = new Date().getTime() - creationDate;
+
+        if (age >= 0 && age < TimeUnit.MINUTES.toMillis(VALID_FOR_MINUTES)) {
+            return new InternalResourceRequest(browserUrl, authenticationName);
+        }
+
+        // too old, so redirect to the real file first
+        return new Redirection(browserUrl);
+    }
+
+    private static class Redirection {
+        private final String url;
+
+        public Redirection(String url) {
+            this.url = url;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            String url = Jenkins.get().getRootUrl() + this.url + restOfPath;
+            rsp.sendRedirect(302, url);
+        }
+    }
+
+    public String getRedirectUrl(Token token, String restOfPath) {
+        String resourceRootUrl = getResourceRootUrl();
+        if (!resourceRootUrl.endsWith("/")) {
+            resourceRootUrl += "/";
+        }
+        if (!restOfPath.startsWith("/")) {
+            // Unsure whether this can happen -- just be safe here
+            restOfPath = "/" + restOfPath;
+        }
+        return resourceRootUrl + getUrlName() + "/" + token.encode() + restOfPath;
```

Comment:
This could potentially be optimized with an explicit `StringBuilder`, though it's possible that current versions of `javac` optimize it the same way.

If you do use `StringBuilder`, then you can update `Token.encode()` to take a `StringBuilder` as a parameter and return `void`.

Reply:
Would consider that premature and expect it to be optimized away anyway.